### PR TITLE
chore: Consolidate prefetching logic within `useTimesForSchedule` and add tests

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

This PR consolidates prefetching logic by moving it from a standalone `usePrefetch` hook into `useTimesForSchedule` as an internal implementation detail. This eliminates prop drilling of `prefetchNextMonth` and `monthCount` through multiple layers of hooks, replacing them with a single optional `bookerLayout` parameter.

**Key Changes:**
- Removes standalone `usePrefetch` hook file
- Simplifies API: replaces 3 parameters (`prefetchNextMonth`, `monthCount`, `month`) with 1 optional parameter (`bookerLayout`)
- Adds comprehensive timezone-aware unit tests (634 lines) covering all layout types and edge cases
- Updates wrapper components to use new simplified API

**Goal:** Improve testability and maintainability by co-locating prefetch logic with the time range calculations it directly affects, without changing any user-facing behavior.

---

## Updates since last revision

**Rebased onto main** to resolve merge conflicts. Conflicts were resolved in:
- `BookerWebWrapper.tsx`: Combined main's `useBookerEmbedEvents` hook with PR's component rename and `bookerLayout` parameter
- `CalendarViewPlatformWrapper.tsx`: Accepted main's refactored structure that delegates to separate `CalendarViewComponent` and `EventTypeCalendarViewComponent`
- `EventTypeCalendarViewComponent.tsx`: Updated to use new `useTimesForSchedule` signature with `bookerLayout` instead of `usePrefetch`

---

**⚠️ Important Notes for Reviewers:**

1. **Three tests are currently skipped** (`it.skip`) with comments indicating they fail during "loading" state - `endTime` is `FEB_END` instead of expected `JAN_END`. These tests are existing failures and I think code should be fixed later.

2. **Component rename**: `BookerPlatformWrapperComponent` → `BookerWebWrapperComponent` in `BookerWebWrapper.tsx` (not mentioned in original description but visible in diff).

3. **Type duplication**: The `bookerLayout` type is defined inline in multiple files. Consider extracting to a shared type for consistency.

Link to Devin run: https://app.devin.ai/sessions/3a78db1b11444f2dbadc967317940242
Requested by: hariom@cal.com (hariom@cal.com) (@hariombalhara)

## Mandatory Tasks

- [ ] I have self-reviewed the code
- [ ] I have updated the developer docs if needed (N/A - internal refactoring only)
- [ ] I confirm automated tests are in place (634 lines of timezone-aware tests added)

## How should this be tested?

**Manual Testing:**
1. Test all booking layouts (month view, week view, column view, mobile)
2. Verify prefetching behavior when:
   - Viewing current month before mid-month (should not prefetch)
   - Viewing current month after mid-month (should prefetch next month)
   - Switching between months
   - Using week view with extra days that cross month boundaries
3. Verify behavior in loading states
4. Test in different timezones to ensure time range calculations remain correct

**Automated Testing:**
```bash
# Run the new test suite
TZ=UTC yarn test useTimesForSchedule.timezone.test.ts
```

**Expected Behavior:**
- Booking availability should be identical to before this PR
- No changes to which time slots are shown to users
- Prefetching should trigger at the same times as before
- All booking layouts should work identically

**Environment:**
- No special environment variables needed
- Tests use mocked store context

## Human Review Checklist

**Critical Items:**
- [ ] Verify conflict resolution in `EventTypeCalendarViewComponent.tsx` - ensure it correctly uses new `useTimesForSchedule` signature
- [ ] Verify `CalendarViewPlatformWrapper.tsx` delegation to separate components works correctly after accepting main's refactoring
- [ ] Investigate the 3 skipped tests - are they bugs or incorrect test expectations?
- [ ] Verify `monthFromStore ?? monthFromProps` priority order is intentional
- [ ] Confirm no user-facing behavior changes by testing booking flows manually
- [ ] Verify behavior when `bookerLayout` is undefined (graceful degradation expected)

**Code Quality:**
- [ ] Consider extracting `bookerLayout` type to shared definition
- [ ] Review TODO comment about `skipToken` for timezone placeholder
- [ ] Verify internal `usePrefetch` properly respects all `bookerState` values

---

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have checked that my changes generate no new warnings
- [x] I have read the contributing guide

### CODE REVIEW

**Comment by hariombalhara** [comment]
* **#23918** <a href="https://app.graphite.com/github/pr/calcom/cal.com/23918?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/calcom/cal.com/23918?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#23913** <a href="https://app.graphite.com/github/pr/calcom/cal.com/23913?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/features/schedules/lib/use-schedule/useSchedule.ts [hunk 1]
```diff
@@ -18,16 +18,19 @@ export type UseScheduleWithCacheArgs = {
...
   isTeamEvent?: boolean;
   orgSlug?: string;
   teamMemberEmail?: string | null;
   useApiV2?: boolean;
   enabled?: boolean;
+  bookerLayout: {
+    layout: string;
+    extraDays: number;
+    columnViewExtraDays: { current: number };
+  };
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useSchedule.ts]: _bookerLayout_ is needed by _usePrefetch_ which is now called internally by useTimesForSchedule

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 2]
```diff
@@ -1,24 +1,110 @@
...
-  monthCount,
   selectedDate,
   dayCount,
-  prefetchNextMonth,
+  bookerLayout,
 }: UseTimesForScheduleProps): [string, string] => {
+  // Get values from booker store
+  const [monthFromStore, bookerState] = useBookerStoreContext((state) => [state.month, state.state], shallow);
+
+  const resolvedMonth = monthFromStore ?? month ?? null;
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: This is how the BookerWebWrapper and BookerPlatformWrapper were resolving month preferring the monthFromStore and falling back to monthFromProperty

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 3]
```diff
@@ -1,24 +1,110 @@
...
+  const { pretchFutureMonths, monthsToPrefetch } = usePrefetch({
+    date,
+    month: resolvedMonth,
+    bookerLayout,
+    bookerState,
+  });
+
   const now = dayjs();
-  const monthDayjs = month ? dayjs(month) : now;
-  const nextMonthDayjs = monthDayjs.add(monthCount ? monthCount : 1, "month");
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:18]: nextMonthDayjs -> lastMonthToPrefetchDayjs because it is not necessarily the next month, it could be any number of months and we get the last month here.

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 4]
```diff
@@ -35,7 +121,7 @@ export const useTimesForSchedule = ({
     }
   } else {
     startTime = monthDayjs.startOf("month").toISOString();
-    endTime = (prefetchNextMonth ? nextMonthDayjs : monthDayjs).endOf("month").toISOString();
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:38]: _prefetchNextMonth_ renamed to _pretchFutureMonths_ because it isn't just next month

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 5]
```diff
@@ -1,24 +1,110 @@
...
+  // TODO: We need to refactor the approach to prefetch and prefetching must be done as a second non-blocking request instead of fetching all data together
+  const monthsToPrefetch: 1 | 2 =
+    // Don't know why we need to specifically prevent prefetching for non-Week view layouts when booker state isn't selecting_time(e.g. loading)
+    // This ends up causing different time ranges for different booker states and thus multiple getSchedule calls are made
+    // Steps to replicate: Simply go to http://acme.cal.local:3000/owner1/30min?overlayCalendar=true&layout=month_view&month=2025-09&date=2025-09-24. Two calls will be made, first one that prefetches 1 month and second one that prefetches 2 months.
+    ((bookerLayout.layout !== BookerLayouts.WEEK_VIEW && bookerState === "selecting_time") ||
+      bookerLayout.layout === BookerLayouts.COLUMN_VIEW) &&
+    doesExtraDaysNeedOneMoreMonth
+      ? 2
+      : 1;
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: _undefined_ replaced with 1 directly because _useTimesForSchedule_ which uses it was doing this anyway. having an _undefined_ value here was weird.

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 6]
```diff
@@ -1,24 +1,110 @@
...
+  // Practically, this conditon doesn't frequently arise because it needs to have date query param set. So, to be able to replicate this issue, one needs to select a date and then refresh the page
+  // Fix it in a follow up PR - There is a failing test that is skipped because of this issue
+  const doesExtraDaysNeedOneMoreMonth =
+    !isNaN(monthAfter1MonthFromDate) &&
+    !isNaN(monthAfterColumnViewExtraDaysFromDate) &&
+    // monthAfterColumnViewExtraDaysFromDate > monthAfter1MonthFromDate &&
+    monthAfterColumnViewExtraDaysFromDate != monthAfter1MonthFromDate;
+
+  // TODO: We need to refactor the approach to prefetch and prefetching must be done as a second non-blocking request instead of fetching all data together
+  const monthsToPrefetch: 1 | 2 =
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: 1 | 2 makes it clear that no other values are possible at the moment

File: packages/platform/atoms/booker/BookerWebWrapper.tsx [hunk 7]
```diff
@@ -37,7 +35,7 @@ export type BookerWebWrapperAtomProps = BookerProps & {
   eventData?: NonNullable<Awaited<ReturnType<typeof getPublicEvent>>>;
 };
 
-const BookerPlatformWrapperComponent = (props: BookerWebWrapperAtomProps) => {
+const BookerWebWrapperComponent = (props: BookerWebWrapperAtomProps) => {
```
Comment (hariombalhara) [inline packages/platform/atoms/booker/BookerWebWrapper.tsx:40]: Wrong name of component

File: packages/platform/atoms/booker/BookerPlatformWrapper.tsx [hunk 8]
```diff
@@ -233,22 +232,14 @@ const BookerPlatformWrapperComponent = (
...
     return restFormValues;
   }, [restFormValues]);
-  const date = dayjs(selectedDate).format("YYYY-MM-DD");
 
-  const { prefetchNextMonth, monthCount } = usePrefetch({
-    date,
-    month,
-    bookerLayout,
-    bookerState,
-  });
```
Comment (hariombalhara) [inline packages/platform/atoms/booker/BookerPlatformWrapper.tsx:247]: Unused

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 9]
```diff
@@ -1,24 +1,106 @@
...
+  month: string | null;
+  bookerLayout?: {
+    layout: string;
+    extraDays: number;
+    columnViewExtraDays: { current: number };
+  };
+  bookerState: BookerState;
+}
+
+export const usePrefetch = ({ date, month, bookerLayout, bookerState }: UsePrefetchParams) => {
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: The _usePrefetch_ hook is moved as is from  _usePrefetch_ file to here except some minor changes like variables renames, additional comments and returning just _monthsToPrefetch_. We don't need a boolean(prefetchFutureMonths) because _monthsToPrefetch_ if null could be considered as a hint to not prefetch at all.

Because usePrefetch is only to be used by _useTimesForSchedule_ hook

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 10]
```diff
@@ -34,8 +116,10 @@ export const useTimesForSchedule = ({
       endTime = monthDayjs.startOf("month").add(dayCount, "day").toISOString();
     }
   } else {
+    const monthsToPrefetch = prefetchData?.monthsToPrefetch;
+    const lastMonthToPrefetchDayjs = monthsToPrefetch ? monthDayjs.add(monthsToPrefetch, "month") : null;
     startTime = monthDayjs.startOf("month").toISOString();
-    endTime = (prefetchNextMonth ? nextMonthDayjs : monthDayjs).endOf("month").toISOString();
+    endTime = (lastMonthToPrefetchDayjs ? lastMonthToPrefetchDayjs : monthDayjs).endOf("month").toISOString();
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:101]: Using _nextMonthDayjs_(which is now renamed to _lastMonthToPrefetchDayjs_) itself to check condition instead of checking a separate boolean and then using it.

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.timezone.test.ts [hunk 11]
```diff
@@ -0,0 +1,622 @@
...
+}));
+
+// Timezone offset mapping (in minutes from UTC)
+const TIMEZONE_OFFSETS = {
+  "Asia/Kolkata": 330, // UTC+5:30
+  "Asia/Calcutta": 330, // Legacy name for Asia/Kolkata
+  UTC: 0, // UTC+0
+  "Etc/UTC": 0, // Alternative UTC
+  GMT: 0, // GMT is effectively UTC
+} as const;
```
[the remark this answers is not in the document]
Comment (hariombalhara) [reply packages/features/schedules/lib/use-schedule/useTimesForSchedule.timezone.test.ts:24]: That is the intention.

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts (hunk 9 above)
Comment (volnei) [reply packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: Can we remove the export then?

**Review by volnei — DISMISSED** [review]

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 12]
```diff
@@ -1,24 +1,106 @@
...
+    // This ends up causing different time ranges for different booker states and thus multiple getSchedule calls are made
+    // Steps to replicate: Simply go to http://acme.cal.local:3000/owner1/30min?overlayCalendar=true&layout=month_view&month=2025-09&date=2025-09-24. Two calls will be made, first one that prefetches 1 month and second one that prefetches 2 months.
+    ((bookerLayout.layout !== BookerLayouts.WEEK_VIEW && bookerState === "selecting_time") ||
+      bookerLayout.layout === BookerLayouts.COLUMN_VIEW) &&
+    doesExtraDaysNeedOneMoreMonth
+      ? 2
+      : 1;
+
+  return { monthsToPrefetch };
+};
```
[the remark this answers is not in the document]
Comment (hariombalhara) [reply packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:62]: Are you sure, I think when we pass null to dayjs(null) it gives 'Invalid Date' and it stays invalid date after all the operations

File: packages/features/schedules/lib/use-schedule/useSchedule.ts [hunk 13]
```diff
@@ -87,6 +91,7 @@ export const useSchedule = ({
     startTime,
     // if `prefetchNextMonth` is true, two months are fetched at once.
     endTime,
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (joeauyeung) [inline packages/features/schedules/lib/use-schedule/useSchedule.ts]: @volnei what's your thought here? I know we want to avoid any new eslint ignore comments. Instead should we be handling if timeZone is null even if it's rarely the case. We'd rather be defensive about this than running into an error in prod.

File: packages/platform/atoms/booker/BookerWebWrapper.tsx (hunk 7 above)
Comment (joeauyeung) [reply packages/platform/atoms/booker/BookerWebWrapper.tsx:40]: Wow good catch with that one

**Review by joeauyeung — DISMISSED** [review]
LGTM, nice refactor. 

As a follow up, I think it would be cool to replace the `dayjs` functions with `date-fns` since we're not concerned about timezones in the `useTimesForSchedule` hook

File: packages/features/schedules/lib/use-schedule/useSchedule.ts (hunk 13 above)
Comment (hariombalhara) [reply packages/features/schedules/lib/use-schedule/useSchedule.ts]: Fixed it

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 14]
```diff
@@ -1,24 +1,106 @@
...
+      !isNaN(dateMonth) &&
+      !isNaN(monthAfterExtraDaysFromDate) &&
+      dateMonth !== monthAfterExtraDaysFromDate) ||
+    (bookerLayout.layout === BookerLayouts.COLUMN_VIEW &&
+      !isNaN(dateMonth) &&
+      !isNaN(monthAfterColumnViewExtraDaysFromDate) &&
+      dateMonth !== monthAfterColumnViewExtraDaysFromDate) ||
+    ((bookerLayout.layout === BookerLayouts.MONTH_VIEW || bookerLayout.layout === "mobile") &&
+      (!isValidDate || isSameMonth) &&
+      isAfter2Weeks);
```
Comment (Udit-takkar) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: This is not easily readble or understandable. we should refactor this 

 ```
 const isValidMonth = (month: number) => !isNaN(month);

  const shouldPrefetchForWeekView = () => {
    if (bookerLayout.layout !== BookerLayouts.WEEK_VIEW || !bookerLayout.extraDays) {
      return false;
    }
    const targetMonth = monthAfterExtraDaysFromDate;
    return isValidMonth(dateMonth) && isValidMonth(targetMonth) && dateMonth !== targetMonth;
  };

  const shouldPrefetchForColumnView = () => {
    if (bookerLayout.layout !== BookerLayouts.COLUMN_VIEW) {
      return false;
    }
    const targetMonth = monthAfterColumnViewExtraDaysFromDate;
    return isValidMonth(dateMonth) && isValidMonth(targetMonth) && dateMonth !== targetMonth;
  };

  const shouldPrefetchForMonthOrMobileView = () => {
    const isMonthOrMobile = bookerLayout.layout === BookerLayouts.MONTH_VIEW || bookerLayout.layout === "mobile";
    if (!isMonthOrMobile) {
      return false;
    }
    return (!isValidDate || isSameMonth) && isAfter2Weeks;
  };
  
   const pretchFutureMonths = 
    shouldPrefetchForWeekView() || 
    shouldPrefetchForColumnView() || 
    shouldPrefetchForMonthOrMobileView();
```

Comment (hariombalhara) [reply packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts]: The goal is to merge it as is with minimal changes required to make it testable. I think we can take care of it in followup

**Comment by Udit-takkar** [comment]
@hariombalhara rajiv also did a refactor recently for usePrefetch. could you fix the conflict?

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 15]
```diff
@@ -1,24 +1,85 @@
...
+  month: string | null;
+  bookerLayout?: {
+    layout: string;
+    extraDays: number;
+    columnViewExtraDays: { current: number };
+  };
+  bookerState: BookerState;
+}
+
+const usePrefetch = ({ date, month, bookerLayout, bookerState }: UsePrefetchParams) => {
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:27]: Moved from _usePrefetch_ module here.

Comment (hariombalhara) [reply packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:27]: bookerLayout param is newly added.

**Comment by hariombalhara** [comment]
@Udit-takkar resolved conflicts

File: packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts [hunk 16]
```diff
@@ -1,24 +1,85 @@
...
-  month,
-  monthCount,
+  month: monthFromProps,
   selectedDate,
   dayCount,
-  prefetchNextMonth,
+  bookerLayout,
 }: UseTimesForScheduleProps): [string, string] => {
+  const [monthFromStore, bookerState] = useBookerStoreContext((state) => [state.month, state.state], shallow);
+  const month = monthFromStore ?? monthFromProps ?? null;
```
Comment (hariombalhara) [inline packages/features/schedules/lib/use-schedule/useTimesForSchedule.ts:71]: Platform Component uses month from store already(see https://github.com/calcom/cal.com/blob/09-18-add-tests-_useprefetch/packages/platform/atoms/booker/BookerPlatformWrapper.tsx#L208)

To keep it consistent I am giving monthFromStore preference. Here also we give the same preference https://github.com/calcom/cal.com/blob/09-18-add-tests-_useprefetch/packages/features/bookings/Booker/utils/event.ts#L113

**Review by Udit-takkar — DISMISSED** [review]

**Comment by Udit-takkar** [comment]
@hariombalhara Can you please fix the conflicts and get review from Platform Team?

**Comment by hariombalhara** [comment]
@Udit-takkar Thanks for reminder, did it.

**Review by volnei — APPROVED** [review]
